src/proxy/http/profiles.rs

```diff
@@ -482,25 +494,37 @@ pub mod router {
         R::Value: svc::Service<http::Request<B>> + Clone,
     {
         fn update_routes(&mut self, routes: Routes) {
-            let slots = routes.len() + 1;
+            let slots = routes.routes.len() + 1;
             self.router = Router::new(
                 Recognize {
                     target: self.target.clone(),
-                    routes,
+                    routes: routes.routes,
                     default_route: self.default_route.clone(),
                 },
                 self.stack.clone(),
                 slots,
                 // Doesn't matter, since we are guaranteed to have enough capacity.
                 Duration::from_secs(0),
             );
+            self.dst_overrides = routes.dst_overrides;
         }
 
         fn poll_route_stream(&mut self) -> Option<Async<Option<Routes>>> {
             self.route_stream
                 .as_mut()
                 .and_then(|ref mut s| s.poll().ok())
         }
+
+        fn pick_dst_override(&self) -> Option<NameAddr> {
+            if !self.dst_overrides.is_empty() {
+                let weights = self.dst_overrides.iter().map(|dst| dst.1);
+                let mut rng = rand::thread_rng();
+                let dist = WeightedIndex::new(weights).unwrap();
+                Some(self.dst_overrides[dist.sample(&mut rng)].0.clone())
```

Comment:
Nit/TIOLI: Personally, I find this expression a little complex and hard to follow. What about something like:

```rust
let idx = WeightedIndex::new(weights)
    .unwrap()
    .sample(&mut rng);
Some(self.dst_overrides[idx].0.clone())
```